main/src/cgeo/geocaching/calculator/FormulaParser.java

```diff
@@ -0,0 +1,379 @@
+
+/**
+ * Parse coordinates with formulas.
+ */
+public final class FormulaParser {
+
+    public  static final char WPC_DELIM = '|';
+
+    private static final Pattern PATTERN_BAD_BLANK_COMMA = Pattern.compile("(\\d), (\\d{2,})");
+    private static final Pattern PATTERN_BAD_BLANK_DOT = Pattern.compile("(\\d)\\. (\\d{2,})");
+
+    private static final List<AbstractFormulaParser> parsers = Arrays.asList(new MinDecFormulaParser());
+
+    private Settings.CoordInputFormatEnum desiredFormulaFormat = null;
+
+    public FormulaParser() {
+    }
+
+    public FormulaParser(final Settings.CoordInputFormatEnum formulaFormat) {
+        desiredFormulaFormat = formulaFormat;
+    }
+
+    public static class ParseException extends IllegalArgumentException {
+        private static final long serialVersionUID = 1L;
+        public final int resource;
+
+        public ParseException(final String msg) {
+            super(msg);
+            resource = R.string.err_parse_lat_lon;
+        }
+
+        public ParseException(final String msg, final Geopoint.LatLon faulty) {
+            super(msg);
+            resource = faulty == Geopoint.LatLon.LAT ? R.string.err_parse_lat : R.string.err_parse_lon;
+        }
+    }
+
+    private static class ResultWrapper {
+        private final String result;
+        private final int matcherLength;
+
+        ResultWrapper(final String result, final int stringLength) {
+            this.result = result;
+            this.matcherLength = stringLength;
+        }
+    }
+
+    /**
+     * Abstract parser for coordinate formats.
+     */
+    private abstract static class AbstractFormulaParser {
+        /**
+         * Parses coordinates (with formula) out of the given string for a specific coordinate format.
+         *
+         * @param text the string to be parsed
+         * @return an pair of strings with parsed formula for latitude and longitude
+         */
+        @Nullable
+        public abstract ImmutablePair<String, String> parse(@NonNull String text);
+
+        /**
+         * Parses latitude or longitude out of the given string.
+         *
+         * @param text
+         *            the string to be parsed
+         * @param latlon
+         *            whether to parse latitude or longitude
+         * @return a wrapper with the parsed latitude/longitude and the length of the match, or null if parsing failed
+         */
+        @Nullable
+        public abstract ResultWrapper parse(@NonNull String text, @NonNull Geopoint.LatLon latlon);
+
+        /**
+         * Indicates which format is parsed
+         * @return coordInputFormat which can be parsed
+         */
+        public abstract Settings.CoordInputFormatEnum formulaFormat();
+    }
+
+    /**
+     * Abstract parser for coordinates that consist of two syntactic parts: latitude and longitude.
+     */
+    private abstract static class AbstractLatLonFormulaParser extends AbstractFormulaParser {
```

Comment:
Is AbstractLatLonFormulaParser the only subclass of AbstractFormularParser? If yes, why do we need the subclass?

Reply:
Probably yes,  but with your previous comment, probably that will be done with reusing the other parser...

Reply:
Did not manage yet to bring to a common base-class or logic...

If I have some more time, I will probably give it another try...